compiler/src/iree/compiler/Dialect/Flow/Transforms/DynamicizeStaticShapes.cpp

```diff
@@ -0,0 +1,278 @@
+    RankedTensorType fullType, RankedTensorType partialType,
+    SmallVectorImpl<OpFoldResult> &offsetValues,
+    SmallVectorImpl<OpFoldResult> &sizeValues) {
+  const SmallVector<OpFoldResult> offsets = op.getMixedOffsets();
+  offsetValues.resize(offsets.size());
+  for (int i = 0, e = offsets.size(); i < e; ++i) {
+    if (auto value = offsets[i].dyn_cast<Value>()) {
+      offsetValues[i] = value;
+      continue;
+    }
+
+    // Create flow.dispatch.dynamicize_dim op for this dimension.
+    auto attr = offsets[i].get<Attribute>().cast<IntegerAttr>();
+    auto dimOp = builder.create<DispatchDynamicizeDimOp>(
+        loc, attr.getValue().getSExtValue());
+    offsetValues[i] = dimOp.getResult();
+  }
+
+  int reducedRank = fullType.getRank() - partialType.getRank();
+  assert(reducedRank >= 0);
+
+  const SmallVector<OpFoldResult> sizes = op.getMixedSizes();
+  sizeValues.resize(sizes.size());
+  for (int i = 0, e = sizes.size(); i < e; ++i) {
+    if (auto value = sizes[i].dyn_cast<Value>()) {
+      sizeValues[i] = value;
+      continue;
+    }
+
+    // Create flow.dispatch.dynamicize_dim op for this dimension.
+    auto attr = sizes[i].get<Attribute>().cast<IntegerAttr>();
+    auto value = attr.getValue().getSExtValue();
+    if (value == 1 && reducedRank > 0) {
+      // For rank reduced dimensions, we need to keep the unit dim as constant
+      // value to make the op semantics sound.
+      sizeValues[i] = attr;
+      --reducedRank;
+    } else {
+      auto dimOp = builder.create<DispatchDynamicizeDimOp>(loc, value);
+      sizeValues[i] = dimOp.getResult();
+    }
+  }
+}
+
+/// Goes through all tensor-shaped ops inside the given `regionOp` and turn
+/// static shapes into dynamic ones.
+LogicalResult dynamicizeStaticShapes(IRRewriter &rewriter,
+                                     DispatchRegionOp &regionOp) {
+  IRRewriter::InsertionGuard guard(rewriter);
+
+  // First get all values defined outside of the dispatch region. Identify
+  // tensor typed values and rewrite them to have dynamic shapes in the dispatch
+  // region.
+  rewriter.setInsertionPoint(regionOp);
+  llvm::SetVector<Value> aboveValues;
+  mlir::getUsedValuesDefinedAbove(regionOp.getBody(), aboveValues);
+  for (int i = 0, e = aboveValues.size(); i < e; ++i) {
+    Value value = aboveValues[i];
+    auto tensorType = value.getType().dyn_cast<RankedTensorType>();
+    if (!tensorType) continue;
+    auto newValue =
+        DispatchDynamicizeShapeOp::build(rewriter, value.getLoc(), value);
+    // Replace all uses in the dispatch region to have dynamic shapes.
+    value.replaceUsesWithIf(newValue, [regionOp](OpOperand &operand) {
+      return operand.getOwner()->getParentOp() == regionOp;
+    });
+  }
+
+  // Then update all compute ops in the dispatch region to return values of
+  // dynamic shapes.
+  Block &block = regionOp.getBody().getBlocks().front();
+  SmallVector<Operation *> tensorOps;
+  for (Operation &op : block.without_terminator()) {
```

Comment:
I think we need a worklist here to transitively follow the uses..... similar to https://github.com/openxla/iree/blob/62ee2f342484a1468abe5bc96a2be7c3192a6e63/compiler/src/iree/compiler/Codegen/Utils/Utils.h#L172